File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-com-5-4/SettlementPeriodsSchedule.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class SettlementPeriodsSchedule : ISerialized
    {
        public SettlementPeriodsSchedule(XmlNode xmlNode)
        {
            XmlNodeList settlementPeriodsStepNodeList = xmlNode.SelectNodes("settlementPeriodsStep");
            
            if (settlementPeriodsStepNodeList != null)
            {
                this.settlementPeriodsStep_ = new List<SettlementPeriodsStep>();
                foreach (XmlNode item in settlementPeriodsStepNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            settlementPeriodsStepIDRef_ = item.Attributes["id"].Value;
                            settlementPeriodsStep_.Add(new SettlementPeriodsStep(item));
                            IDManager.SetID(settlementPeriodsStepIDRef_, settlementPeriodsStep_[settlementPeriodsStep_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            settlementPeriodsStepIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        settlementPeriodsStep_.Add(new SettlementPeriodsStep(item));
                        }
                    }
                    else
                    {
                        settlementPeriodsStep_.Add(new SettlementPeriodsStep(item));
                    }
                }
            }
            
        
            XmlNode deliveryPeriodsReferenceNode = xmlNode.SelectSingleNode("deliveryPeriodsReference");
            
            if (deliveryPeriodsReferenceNode != null)
            {
                if (deliveryPeriodsReferenceNode.Attributes["href"] != null || deliveryPeriodsReferenceNode.Attributes["id"] != null) 
                {
                    if (deliveryPeriodsReferenceNode.Attributes["id"] != null) 
                    {
                        deliveryPeriodsReferenceIDRef_ = deliveryPeriodsReferenceNode.Attributes["id"].Value;
                        CalculationPeriodsReference ob = new CalculationPeriodsReference(deliveryPeriodsReferenceNode);
                        IDManager.SetID(deliveryPeriodsReferenceIDRef_, ob);
                    }
                    else if (deliveryPeriodsReferenceNode.Attributes["href"] != null)
                    {
                        deliveryPeriodsReferenceIDRef_ = deliveryPeriodsReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        deliveryPeriodsReference_ = new CalculationPeriodsReference(deliveryPeriodsReferenceNode);
                    }
                }
                else
                {
                    deliveryPeriodsReference_ = new CalculationPeriodsReference(deliveryPeriodsReferenceNode);
                }
            }
            
        
            XmlNode deliveryPeriodsScheduleReferenceNode = xmlNode.SelectSingleNode("deliveryPeriodsScheduleReference");
            
            if (deliveryPeriodsScheduleReferenceNode != null)
            {
                if (deliveryPeriodsScheduleReferenceNode.Attributes["href"] != null || deliveryPeriodsScheduleReferenceNode.Attributes["id"] != null) 
                {
                    if (deliveryPeriodsScheduleReferenceNode.Attributes["id"] != null) 
                    {
                        deliveryPeriodsScheduleReferenceIDRef_ = deliveryPeriodsScheduleReferenceNode.Attributes["id"].Value;
                        CalculationPeriodsScheduleReference ob = new CalculationPeriodsScheduleReference(deliveryPeriodsScheduleReferenceNode);
                        IDManager.SetID(deliveryPeriodsScheduleReferenceIDRef_, ob);
                    }
                    else if (deliveryPeriodsScheduleReferenceNode.Attributes["href"] != null)
                    {
                        deliveryPeriodsScheduleReferenceIDRef_ = deliveryPeriodsScheduleReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        deliveryPeriodsScheduleReference_ = new CalculationPeriodsScheduleReference(deliveryPeriodsScheduleReferenceNode);
                    }
                }
                else
                {
                    deliveryPeriodsScheduleReference_ = new CalculationPeriodsScheduleReference(deliveryPeriodsScheduleReferenceNode);
                }
            }
            
        
        }
        
    
        #region SettlementPeriodsStep_
        private List<SettlementPeriodsStep> settlementPeriodsStep_;
        public List<SettlementPeriodsStep> SettlementPeriodsStep_
        {
            get
            {
                if (this.settlementPeriodsStep_ != null)
                {
                    return this.settlementPeriodsStep_; 
                }
                else if (this.settlementPeriodsStepIDRef_ != null)
                {
                    return this.settlementPeriodsStep_; 
                }
                else
                {
                      return this.settlementPeriodsStep_; 
                }
            }
            set
            {
                if (this.settlementPeriodsStep_ != value)
                {
                    this.settlementPeriodsStep_ = value;
                }
            }
        }
        #endregion
        
        public string settlementPeriodsStepIDRef_ { get; set; }
        #region DeliveryPeriodsReference_
        private CalculationPeriodsReference deliveryPeriodsReference_;
        public CalculationPeriodsReference DeliveryPeriodsReference_
        {
            get
            {
                if (this.deliveryPeriodsReference_ != null)
                {
                    return this.deliveryPeriodsReference_; 
                }
                else if (this.deliveryPeriodsReferenceIDRef_ != null)
                {
                    deliveryPeriodsReference_ = IDManager.getID(deliveryPeriodsReferenceIDRef_) as CalculationPeriodsReference;
                    return this.deliveryPeriodsReference_; 
                }
                else
                {
                      return this.deliveryPeriodsReference_; 
                }
            }
            set
            {
                if (this.deliveryPeriodsReference_ != value)
                {
                    this.deliveryPeriodsReference_ = value;
                }
            }
        }
        #endregion
        
        public string deliveryPeriodsReferenceIDRef_ { get; set; }
        #region DeliveryPeriodsScheduleReference_
        private CalculationPeriodsScheduleReference deliveryPeriodsScheduleReference_;
        public CalculationPeriodsScheduleReference DeliveryPeriodsScheduleReference_
        {
            get
            {
                if (this.deliveryPeriodsScheduleReference_ != null)
                {
                    return this.deliveryPeriodsScheduleReference_; 
                }
                else if (this.deliveryPeriodsScheduleReferenceIDRef_ != null)
                {
                    deliveryPeriodsScheduleReference_ = IDManager.getID(deliveryPeriodsScheduleReferenceIDRef_) as CalculationPeriodsScheduleReference;
                    return this.deliveryPeriodsScheduleReference_; 
                }
                else
                {
                      return this.deliveryPeriodsScheduleReference_; 
                }
            }
            set
            {
                if (this.deliveryPeriodsScheduleReference_ != value)
                {
                    this.deliveryPeriodsScheduleReference_ = value;
                }
            }
        }
        #endregion
        
        public string deliveryPeriodsScheduleReferenceIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-com-5-4/SettlementPeriodsSchedule.cpp
// SettlementPeriodsSchedule.cpp 
#include "SettlementPeriodsSchedule.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

SettlementPeriodsSchedule::SettlementPeriodsSchedule(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //settlementPeriodsStepNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* settlementPeriodsStepNode = xmlNode->FirstChildElement("settlementPeriodsStep");

   if(settlementPeriodsStepNode){settlementPeriodsStepIsNull_ = false;}
   else{settlementPeriodsStepIsNull_ = true;}

   if(settlementPeriodsStepNode)
   {
      for(settlementPeriodsStepNode; settlementPeriodsStepNode; settlementPeriodsStepNode = settlementPeriodsStepNode->NextSiblingElement("settlementPeriodsStep")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- settlementPeriodsStepNode , address : " << settlementPeriodsStepNode << std::endl;
          #endif
          if(settlementPeriodsStepNode->Attribute("href") || settlementPeriodsStepNode->Attribute("id"))
          {
              if(settlementPeriodsStepNode->Attribute("id"))
              {
                  settlementPeriodsStepIDRef_ = settlementPeriodsStepNode->Attribute("id");
                  settlementPeriodsStep_.push_back(boost::shared_ptr<SettlementPeriodsStep>(new SettlementPeriodsStep(settlementPeriodsStepNode)));
                  settlementPeriodsStep_.back()->setID(settlementPeriodsStepIDRef_);
                  IDManager::instance().setID(settlementPeriodsStepIDRef_, settlementPeriodsStep_.back());
              }
              else if(settlementPeriodsStepNode->Attribute("href")) { settlementPeriodsStepIDRef_ = settlementPeriodsStepNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { settlementPeriodsStep_.push_back(boost::shared_ptr<SettlementPeriodsStep>(new SettlementPeriodsStep(settlementPeriodsStepNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- settlementPeriodsStepNode , address : " << settlementPeriodsStepNode << std::endl;
       #endif
   }

   //deliveryPeriodsReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* deliveryPeriodsReferenceNode = xmlNode->FirstChildElement("deliveryPeriodsReference");

   if(deliveryPeriodsReferenceNode){deliveryPeriodsReferenceIsNull_ = false;}
   else{deliveryPeriodsReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- deliveryPeriodsReferenceNode , address : " << deliveryPeriodsReferenceNode << std::endl;
   #endif
   if(deliveryPeriodsReferenceNode)
   {
      if(deliveryPeriodsReferenceNode->Attribute("href") || deliveryPeriodsReferenceNode->Attribute("id"))
      {
          if(deliveryPeriodsReferenceNode->Attribute("id"))
          {
             deliveryPeriodsReferenceIDRef_ = deliveryPeriodsReferenceNode->Attribute("id");
             deliveryPeriodsReference_ = boost::shared_ptr<CalculationPeriodsReference>(new CalculationPeriodsReference(deliveryPeriodsReferenceNode));
             deliveryPeriodsReference_->setID(deliveryPeriodsReferenceIDRef_);
             IDManager::instance().setID(deliveryPeriodsReferenceIDRef_,deliveryPeriodsReference_);
          }
          else if(deliveryPeriodsReferenceNode->Attribute("href")) { deliveryPeriodsReferenceIDRef_ = deliveryPeriodsReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { deliveryPeriodsReference_ = boost::shared_ptr<CalculationPeriodsReference>(new CalculationPeriodsReference(deliveryPeriodsReferenceNode));}
   }

   //deliveryPeriodsScheduleReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* deliveryPeriodsScheduleReferenceNode = xmlNode->FirstChildElement("deliveryPeriodsScheduleReference");

   if(deliveryPeriodsScheduleReferenceNode){deliveryPeriodsScheduleReferenceIsNull_ = false;}
   else{deliveryPeriodsScheduleReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- deliveryPeriodsScheduleReferenceNode , address : " << deliveryPeriodsScheduleReferenceNode << std::endl;
   #endif
   if(deliveryPeriodsScheduleReferenceNode)
   {
      if(deliveryPeriodsScheduleReferenceNode->Attribute("href") || deliveryPeriodsScheduleReferenceNode->Attribute("id"))
      {
          if(deliveryPeriodsScheduleReferenceNode->Attribute("id"))
          {
             deliveryPeriodsScheduleReferenceIDRef_ = deliveryPeriodsScheduleReferenceNode->Attribute("id");
             deliveryPeriodsScheduleReference_ = boost::shared_ptr<CalculationPeriodsScheduleReference>(new CalculationPeriodsScheduleReference(deliveryPeriodsScheduleReferenceNode));
             deliveryPeriodsScheduleReference_->setID(deliveryPeriodsScheduleReferenceIDRef_);
             IDManager::instance().setID(deliveryPeriodsScheduleReferenceIDRef_,deliveryPeriodsScheduleReference_);
          }
          else if(deliveryPeriodsScheduleReferenceNode->Attribute("href")) { deliveryPeriodsScheduleReferenceIDRef_ = deliveryPeriodsScheduleReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { deliveryPeriodsScheduleReference_ = boost::shared_ptr<CalculationPeriodsScheduleReference>(new CalculationPeriodsScheduleReference(deliveryPeriodsScheduleReferenceNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
std::vector<boost::shared_ptr<SettlementPeriodsStep>> SettlementPeriodsSchedule::getSettlementPeriodsStep()
{
   if(!this->settlementPeriodsStepIsNull_){
        if(settlementPeriodsStepIDRef_ != ""){
             return this->settlementPeriodsStep_;
        }else{
             return this->settlementPeriodsStep_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<SettlementPeriodsStep>>();
   }
}
boost::shared_ptr<CalculationPeriodsReference> SettlementPeriodsSchedule::getDeliveryPeriodsReference()
{
   if(!this->deliveryPeriodsReferenceIsNull_){
        if(deliveryPeriodsReferenceIDRef_ != ""){
             return boost::shared_static_cast<CalculationPeriodsReference>(IDManager::instance().getID(deliveryPeriodsReferenceIDRef_));
        }else{
             return this->deliveryPeriodsReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CalculationPeriodsReference>();
   }
}
boost::shared_ptr<CalculationPeriodsScheduleReference> SettlementPeriodsSchedule::getDeliveryPeriodsScheduleReference()
{
   if(!this->deliveryPeriodsScheduleReferenceIsNull_){
        if(deliveryPeriodsScheduleReferenceIDRef_ != ""){
             return boost::shared_static_cast<CalculationPeriodsScheduleReference>(IDManager::instance().getID(deliveryPeriodsScheduleReferenceIDRef_));
        }else{
             return this->deliveryPeriodsScheduleReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CalculationPeriodsScheduleReference>();
   }
}
}

